src/inmanta/server/services/compilerservice.py

```diff
@@ -477,27 +478,39 @@ async def is_compiling(self, environment_id: uuid.UUID) -> Apireturn:
             return 200
         return 204
 
-    async def _run(self, compile: data.Compile) -> None:
-        """
-        Runs a compile request. At completion, looks for similar compile requests based on _compile_merge_key and marks
-        those as completed as well.
-        """
-        now = datetime.datetime.now()
-
+    async def _calculate_recompile_wait(self, compile: data.Compile):
+        last_run = await data.Compile.get_last_run(compile.environment)
         wait_time = opt.server_autrecompile_wait.get()
-
-        lastrun = await data.Compile.get_last_run(compile.environment)
-        if not lastrun:
+        if not last_run or wait_time == 0:
             wait: float = 0
         else:
-            assert lastrun.completed is not None
-            wait = max(0, wait_time - (now - lastrun.completed).total_seconds())
+            assert last_run.completed is not None
+            if last_run.completed >= compile.requested:
+                wait = max(0, wait_time - (last_run.completed - compile.requested).total_seconds())
```

Comment:
This looks wrong
(I may have messed up the ticket)

# cases 

## no queue for a while
 - wait time 2 s
 - compile was requested at time 7. 
 - last compile ended at time 3. 
 - now is 7.1

We should not wait, last compile is 4 seconds older than request

`last_run.completed < compile.requested`, no wait, correct

## no queue for very short time
 - wait time 2 s
 - compile was requested at time 4. 
 - last compile ended at time 3. 
 - now is 4.1

We should wait, as we are within  2 s

`last_run.completed < compile.requested`, no wait, incorrect

## Reverse, long queue 
 - wait time 2 s
 - compile was requested at time 3. 
 - last compile ended at time 7. 

We should not wait either, request is sufficiently old

2-(7-3) => -2 ==> 0

correct

# improvement

we should wait when 
 1. the request is recent (old condition)
 2. a compile has just complete (new condition)